src/dataflow/src/server.rs

```diff
@@ -218,19 +217,216 @@ pub fn serve(config: Config) -> Result<WorkerGuards<()>, String> {
+    /// Indicates the lowest timestamp across all partitions that we retain bindings for.
+    /// This frontier can be held back by other entities holding the shared
+    /// `TimestampBindingRc`.
+    compaction_frontier: MutableAntichain<Timestamp>,
+}
+
+impl TimestampBindingBox {
+    /// Create a new instance of `TimestampBindingBox`.
+    fn new() -> Self {
+        Self {
+            partitions: HashMap::new(),
+            compaction_frontier: MutableAntichain::new(),
+        }
+    }
+
+    fn adjust_compaction_frontier(
+        &mut self,
+        remove: AntichainRef<Timestamp>,
+        add: AntichainRef<Timestamp>,
+    ) {
+        self.compaction_frontier
+            .update_iter(remove.iter().map(|t| (*t, -1)));
+        self.compaction_frontier
+            .update_iter(add.iter().map(|t| (*t, 1)));
+    }
+
+    /// Drop all of the timestamp bindings for timestamps not in advance of the
+    /// compaction frontier.
+    fn compact(&mut self) {
+        let frontier = self.compaction_frontier.frontier();
+
+        if !frontier.is_empty() {
+            // Other folks are using these timestamp bindings. Lets keep everything
+            // >= what people are using
+            for (_, entries) in self.partitions.iter_mut() {
+                entries.retain(|(time, _)| frontier.less_equal(time));
+            }
+        }
+    }
+
+    /// Add a new mapping from `(partition, offset) -> timestamp`.
+    ///
+    /// Note that the `offset` and `timestamp` both have to be greater than the
+    /// largest previously bound offsets or timestamps for that partition.
+    fn add_binding(&mut self, partition: PartitionId, timestamp: Timestamp, offset: MzOffset) {
+        let entry = self.partitions.entry(partition).or_insert_with(Vec::new);
+        let (last_ts, last_offset) = entry.last().unwrap_or(&(0, MzOffset { offset: 0 }));
+        assert!(
+            offset >= *last_offset,
+            "offset should not go backwards, but {} < {}",
+            offset,
+            last_offset
+        );
+        assert!(
+            timestamp > *last_ts,
+            "timestamp should move forwards, but {} <= {}",
+            timestamp,
+            last_ts
+        );
+        entry.push((timestamp, offset));
+    }
+
+    /// Get the timestamp assignment for `(partition, offset)` if it is known.
+    ///
+    /// This function returns the timestamp and the maximum offset for which it is
+    /// valid.
+    fn get_binding(
+        &self,
+        partition: &PartitionId,
+        offset: MzOffset,
+    ) -> Option<(Timestamp, MzOffset)> {
+        if !self.partitions.contains_key(partition) {
+            return None;
+        }
+
+        let entries = self.partitions.get(partition).expect("known to exist");
+
+        for (ts, max_offset) in entries {
+            if offset <= *max_offset {
+                return Some((*ts, *max_offset));
+            }
+        }
```

Comment:
This looks like it could be done with binary search. Are there any performance considerations / `TODO` comments to jot down about that?

Reply:
added a bsearch todo